### PR TITLE
Support for SentenceTransformers with `deepsparse.sentence_transformers.SentenceTransformer`

### PR DESCRIPTION
**Pull request by mgoin**
Direct replacement for [SentenceTransformer](https://www.sbert.net/#installation) implemented using DeepSparse and optimum-deepsparse

Performance optimizations with bucketing/batching will come in future work

# Smoke Test:
```
pip install -e ".[sentence_transformers]"
```

```python
# from sentence_transformers import SentenceTransformer
from deepsparse.sentence_transformers import SentenceTransformer

model = SentenceTransformer("sentence-transformers/all-MiniLM-L6-v2", export=True)

# Sentences we'd like to encode
sentences = ['This framework generates embeddings for each input sentence',
    'Sentences are passed as a list of string.',
    'The quick brown fox jumps over the lazy dog.']

# Sentences are encoded by calling model.encode()
embeddings = model.encode(sentences)

# Print the embeddings
for sentence, embedding in zip(sentences, embeddings):
    print("Sentence:", sentence)
    print("Embedding:", embedding.shape)
    print("")

### Sentence: This framework generates embeddings for each input sentence
### Embedding: (384,)
### 
### Sentence: Sentences are passed as a list of string.
### Embedding: (384,)
### 
### Sentence: The quick brown fox jumps over the lazy dog.
### Embedding: (384,)
```

# Full test through evaluation with [MTEB](https://github.com/embeddings-benchmark/mteb)
```
pip install mteb deepsparse-nightly[sentence_transformers] sentence-transformers
```
```python
from mteb import MTEB

model_name = "TaylorAI/bge-micro-v2"

from deepsparse.sentence_transformers import SentenceTransformer
model = SentenceTransformer(model_name, export=True)
evaluation = MTEB(tasks=["Banking77Classification"])
results_ds = evaluation.run(model, output_folder=f"results/ds-{model_name}")
print(results_ds)

import sentence_transformers
model = sentence_transformers.SentenceTransformer(model_name)
evaluation = MTEB(tasks=["Banking77Classification"])
results_st = evaluation.run(model, output_folder=f"results/st-{model_name}")
print(results_st)
```

Output:
```
{'Banking77Classification': {'mteb_version': '1.1.1', 'dataset_revision': '0fd18e25b25c072e09e0d92ab615fda904d66300', 'mteb_dataset_name': 'Banking77Classification', 'test': {'accuracy': 0.8117207792207791, 'f1': 0.8109893836310513, 'accuracy_stderr': 0.007164150669501205, 'f1_stderr': 0.007346045502756079, 'main_score': 0.8117207792207791, 'evaluation_time': 26.05}}}

{'Banking77Classification': {'mteb_version': '1.1.1', 'dataset_revision': '0fd18e25b25c072e09e0d92ab615fda904d66300', 'mteb_dataset_name': 'Banking77Classification', 'test': {'accuracy': 0.8117207792207791, 'f1': 0.8109893836310513, 'accuracy_stderr': 0.007164150669501205, 'f1_stderr': 0.007346045502756079, 'main_score': 0.8117207792207791, 'evaluation_time': 12.21}}}
```

### CODE REVIEW

File: src/deepsparse/sentence_transformers/sentence_transformer.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        input_was_string = False
+        if isinstance(sentences, str) or not hasattr(
+            sentences, "__len__"
+        ):  # Cast an individual sentence to a list with length 1
+            sentences = [sentences]
+            input_was_string = True
+
+        all_embeddings = []
+        length_sorted_idx = np.argsort([-self._text_length(sen) for sen in sentences])
+        sentences_sorted = [sentences[idx] for idx in length_sorted_idx]
```
Comment (bfineran) [inline src/deepsparse/sentence_transformers/sentence_transformer.py:84]: could directly do this sort with `sort(..., key=...`)

File: src/deepsparse/sentence_transformers/sentence_transformer.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            )
+
+            if output_value == "token_embeddings":
+                embeddings = []
+                for token_emb, attention in zip(
+                    out_features[output_value], out_features["attention_mask"]
+                ):
+                    last_mask_id = len(attention) - 1
+                    while last_mask_id > 0 and attention[last_mask_id].item() == 0:
+                        last_mask_id -= 1
```
Comment (bfineran) [inline src/deepsparse/sentence_transformers/sentence_transformer.py:106]: is attention mask gauranteed to be 1s followed by 0s? could remove this while loop with attention.sum() and a slice, otherwise, could select from `token_emb` by `attention.nonzero()`

File: src/deepsparse/sentence_transformers/sentence_transformer.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            out_features["sentence_embedding"] = self.mean_pooling(
+                model_output, model_inputs["attention_mask"]
+            )
+
+            if output_value == "token_embeddings":
+                embeddings = []
+                for token_emb, attention in zip(
+                    out_features[output_value], out_features["attention_mask"]
+                ):
+                    last_mask_id = len(attention) - 1
```
Comment (bfineran) [inline src/deepsparse/sentence_transformers/sentence_transformer.py:104]: nit: quick in-line comment to say this is to apply the mask

File: src/deepsparse/sentence_transformers/sentence_transformer.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    def encode(
+        self,
+        sentences: Union[str, List[str]],
+        batch_size: int = 1,
+        show_progress_bar: bool = None,
+        output_value: str = "sentence_embedding",
+        convert_to_numpy: bool = True,
+        convert_to_tensor: bool = False,
+        normalize_embeddings: bool = False,
+    ) -> Union[List[torch.Tensor], np.ndarray, torch.Tensor]:
```
Comment (bfineran) [inline src/deepsparse/sentence_transformers/sentence_transformer.py:57]: missing docstring

**Review by bfineran — COMMENTED** [review]
LGTM overall - assuming a readme will come separately.  will fit in nicely with the pipelines refactor - should be able to swap in `optimum.deepsparse` the way we swap in any other engine to our pipelines

File: src/deepsparse/sentence_transformers/sentence_transformer.py (hunk 2 above)
Comment (mgoin) [reply src/deepsparse/sentence_transformers/sentence_transformer.py:106]: This part was copied from SentenceTransformers, so good call on simplifying this. I believe in the guarantee so I will choose the first implementation

File: setup.py [hunk 5]
```diff
@@ -151,6 +152,9 @@ def _parse_requirements_file(file_path):
     "scikit-learn",
     "seqeval",
 ]
+_sentence_transformers_integration_deps = [
+    "optimum-deepsparse @ git://github.com/neuralmagic/optimum-deepsparse.git"
+] + _torch_deps
```
Comment (bfineran) [inline setup.py:157]: @dhuangnm will this build into a pypi wheel? Think it may not

**Review by bfineran — DISMISSED** [review]
LGTM pending coment

File: setup.py (hunk 5 above)
Comment (dhuangnm) [reply setup.py:157]: I think we'll probably need to build optimum-deepsparse wheel separately as we do for nm-transformers and nm-yolov5, then push it out to public pypi in order to install from it?

Comment (bfineran) [reply setup.py:157]: if we want `pip install deepsparse[sentence-transformers]` to work from pypi this is probably an option we'll have to consider
